src/mon/OSDMonitor.cc

```diff
@@ -7784,22 +7784,30 @@ int OSDMonitor::get_crush_rule(const string &rule_name,
   return 0;
 }
 
-int OSDMonitor::check_pg_num(int64_t pool, int pg_num, int size, ostream *ss)
+int OSDMonitor::check_pg_num(int64_t pool, int pg_num, int size, int crush_rule, ostream *ss)
 {
   auto max_pgs_per_osd = g_conf().get_val<uint64_t>("mon_max_pg_per_osd");
-  auto num_osds = std::max(osdmap.get_num_in_osds(), 3u);   // assume min cluster size 3
-  auto max_pgs = max_pgs_per_osd * num_osds;
   uint64_t projected = 0;
-  if (pool < 0) {
-    projected += pg_num * size;
-  }
-  for (const auto& i : osdmap.get_pools()) {
-    if (i.first == pool) {
-      projected += pg_num * size;
-    } else {
-      projected += i.second.get_pg_num_target() * i.second.get_size();
+  unsigned osd_num = 0;
+  auto num_osds = std::max(osdmap.get_num_in_osds(), 3u);
+  CrushWrapper newcrush;
+  if (mapping.get_epoch() >= osdmap.get_epoch()) {
+    _get_pending_crush(newcrush);
+    set<int> roots;
+    newcrush.find_takes_by_rule(crush_rule, &roots);
+    if (!roots.empty()) {
```

Comment:
could you reconsider the comment of https://github.com/ceph/ceph/pull/39062#discussion_r619601489 ?

Reply:
Sorry, I deleted it because I thought the previous treatment was a bit redundant, I will change it back again.

Reply:
i am quoting my comment as a reference

> nit, i'd avoid make an empty `osd_ids` a special case, as the following code block is able to handle this case just fine, and it does not incur a significant performance penalty.

in other words, i suggest remove the `if (!roots.empty()) {` check.

Reply:
@fyzard1991 this comment is not addressed. could you reconsider it?